zb-db/src/test/java/io/camunda/zeebe/db/impl/ColumnFamilyPropertyTest.java

```diff
@@ -0,0 +1,205 @@
+import org.assertj.core.util.Files;
+
+public class ColumnFamilyPropertyTest {
+
+  private Map<Long, Long> map;
+  private ColumnFamily<DbLong, DbLong> columnFamily;
+
+  @Property
+  void columnFamilyBehavesLikeMap(
+      @ForAll("operations") final Iterable<TestableOperation> operations) {
+    operations.forEach(op -> op.runOnMap(map));
+    operations.forEach(op -> op.runOnColumnFamily(columnFamily));
+    columnFamilyBehavedLikeMap(columnFamily, map);
+  }
+
+  private void columnFamilyBehavedLikeMap(
+      final ColumnFamily<DbLong, DbLong> columnFamily, final Map<Long, Long> map) {
+    map.forEach(
+        (key, value) -> {
+          final var dbKey = new DbLong();
+          dbKey.wrapLong(key);
+          assertThat(columnFamily.get(dbKey).getValue())
+              .as("Key " + dbKey.getValue() + " should have value " + value)
+              .isEqualTo(value);
+        });
+    columnFamily.forEach(
+        (key, value) -> assertThat(map).containsEntry(key.getValue(), value.getValue()));
+  }
+
+  @Provide
+  ListArbitrary<TestableOperation> operations() {
+    final var k = Arbitraries.longs().greaterOrEqual(0);
+    final var v = Arbitraries.longs();
+    final var insert =
+        Combinators.combine(k, v)
+            .as(
+                (arbitraryK, arbitraryV) ->
+                    (TestableOperation) new InsertOp(arbitraryK, arbitraryV));
+    final var update =
+        Combinators.combine(k, v)
+            .as(
+                (arbitraryK, arbitraryV) ->
+                    (TestableOperation) new UpdateOp(arbitraryK, arbitraryV));
+    final var upsert =
+        Combinators.combine(k, v)
+            .as(
+                (arbitraryK, arbitraryV) ->
+                    (TestableOperation) new UpsertOp(arbitraryK, arbitraryV));
+    final var deleteExisting =
+        Combinators.combine(k, v)
+            .as(
+                (arbitraryK, arbitraryV) ->
+                    (TestableOperation) new DeleteExistingOp(arbitraryK, arbitraryV));
+    final var c =
+        Combinators.combine(k, v)
+            .as(
+                (arbitraryK, arbitraryV) ->
+                    (TestableOperation) new DeleteIfExistsOp(arbitraryK, arbitraryV));
+
+    return Arbitraries.oneOf(List.of(insert, update, upsert, deleteExisting)).list();
+  }
+
+  @BeforeProperty
+  private void setup() {
+    final File pathName = Files.newTemporaryFolder();
+    final ZeebeDbFactory<DefaultColumnFamily> dbFactory = DefaultZeebeDbFactory.getDefaultFactory();
+    final ZeebeDb<DefaultColumnFamily> zeebeDb = dbFactory.createDb(pathName);
+
+    columnFamily =
+        zeebeDb.createColumnFamily(
+            DefaultColumnFamily.DEFAULT, zeebeDb.createContext(), new DbLong(), new DbLong());
+    map = new HashMap<>();
+  }
+
+  @AfterTry
+  private void cleanup() {
+    columnFamily.forEach((k, v) -> columnFamily.delete(k));
+    map.clear();
+  }
+
+  record InsertOp(long k, long v) implements TestableOperation {
```

Comment:
:wrench: One way to save a lot of code would be something like this:

```
  @FunctionalInterface
  interface Modification<STATE, KEY, VALUE> {
    void apply(STATE state, KEY ke, VALUE value);
  }

  enum Op {
    INSERT((t, k, v) -> t.insert(k, v), (m,k,v) -> m.put(k,v)),
    DELETE_EXISTING((t, k, v) -> t.deleteExisting(k), (m,k,v) -> m.remove(k))
   ...


    private final Modification<ColumnFamily<DbLong, DbLong>, DbLong, DbLong> tableModification;
    private final Modification<Map<Long, Long>, Long, Long> mapModification;

    Op(final Modification<ColumnFamily<DbLong, DbLong>, DbLong, DbLong> tableModification,
        final Modification<Map<Long, Long>, Long, Long> mapModification) {
      this.tableModification = tableModification;
      this.mapModification = mapModification;
    }


    public void runOnColumnFamily(final ColumnFamily<DbLong, DbLong> columnFamily, final long key, final long value) {
      final var dbKey = new DbLong();
      final var dbValue = new DbLong();
      dbKey.wrapLong(key);
      dbValue.wrapLong(value);
      try {
        tableModification.apply(columnFamily, dbKey, dbValue);
      } catch (final ZeebeDbInconsistentException ignored) {
        // Ignore failed insert due to duplicates
      }
    }

    public void runOnMap(final Map<Long, Long> map, final long key, final long value) {
      mapModification.apply(map, key, value);
    }
  }
```

And then you could generate an arbitrary of a list of enums, which could also simplify some of the code above.

Reply:
I've used a slightly different approach that's also a bit less redundant but might allow for further extension to `forEach` operations etc. Let me know what you think :)